### PR TITLE
[Refs #346] Named ratio/crop modifiers

### PR DESCRIPTION
**Pull request by herzinger**

### CODE REVIEW

**Review by florianbouvot — CHANGES_REQUESTED** [review]
1 - Can we provide a consistent default crop/ratio list ?
Like `3:1`; `2:1`, `16:9`, `4:3` most common video and photo aspect ratio.

2 - Can we harmonize default naming in map ?

Like:
```
$inuit-crops|ratios: (
  "3\\:1" : (3:1),
  "2\\:1" : (2:1),
  "16\\:9" : (16:9),
  "4\\:3" : (4:3),
  "golden-ratio" : (1.618:1) // Not needed by default
) !default;
```

And add description about supported naming `golden-ratio`...?

**Comment by herzinger** [comment]
I thought about both, and decided against, for one reason: the user shouldn't really use this map anyway. No real project will use the exact same ratios we define here, this is more of a showcase than an actual, useful list. That being the case, the map I defined is a good demo of variated use cases. The different names show some conventions and how to achieve then (and I'd argue the best one should be to give a project specific name, like "hero" or "card"), and the golden ratio shows you can use non-integers. 

To be honest, my preferred solution would be to define it as `null`, like what we decided with the responsive spacings, and have this map as an example in a comment, but I decided to leave it as a discussion for another time, to be something of a framework-wide mentality, maybe in a future version.

**Comment by herzinger** [comment]
Another point: would we stick with any one convention, it would look like we are endorsing this specific usage. And since the best practice is, in my opinion, to define something project specific, I believe nothing we could write here would constitute good advice, really.

Of course, nothing I said here is out of discussion, and I can change it if the team decide I should.

**Review by csshugs — APPROVED** [review]
> I believe nothing we could write here would constitute good advice, really.

I'm with @herzinger here. I consider it a sensible idea to provide all the possible ways of naming a class and let the user override the default ratios anyway.

**Comment by dennisfrank** [comment]
@herzinger There is [wrong class in line 131](https://github.com/herzinger/inuitcss/blob/ae258bc0dccacebc8a1b3cfb50030c92fd5e58ad/objects/_objects.crop.scss#L131):

`.o-ratio--#{$crop-name}` instead of `.o-crop--#{$crop-name}`

File: objects/_objects.crop.scss [hunk 1]
```diff
@@ -103,66 +106,36 @@ $inuit-crops: (
...
-      $consequent-integer: str-slice(inspect($consequent), 1, str-index(inspect($consequent), ".") - 1);
-      // Get everything after the “.”
-      $consequent-decimal: str-slice(inspect($consequent), str-index(inspect($consequent), ".") + 1);
-      // Put the number back together with a “\”-prefixed decimal separator.
-      $consequent-class: $consequent-integer + "\\." + $consequent-decimal;
+      @error "`#{$consequent}` needs to be a number.";
     }
 
-    .o-crop--#{$antecedent-class}\:#{$consequent-class} {
+    .o-ratio--#{$crop-name} {
```
Comment (nenadjelovac) [inline objects/_objects.crop.scss]: As mentioned in https://github.com/inuitcss/inuitcss/pull/352#issuecomment-367112798, this class should be `.o-ratio--#{$crop-name}`

Comment (nenadjelovac) [reply objects/_objects.crop.scss]: @dennisfrank thnx for catching this ;) It's getting late in my timezone, the brain is out :D

**Comment by herzinger** [comment]
@dennisfrank thanks, it should be right now.

Besides that, what is the consensus on the provided default map for this pull request in the end?

**Comment by csshugs** [comment]
@herzinger As I already stated, I'm fine with it as it showcases the variety of possible ways to name the classes. This is self-documented code par excellence. It's likely that the inuitcss user overrides the map anyway which is easy to do in case.

@florianbouvot Do you have any points against the proposed default maps, that we didn't discussed yet?

**Comment by nenadjelovac** [comment]
@csshugs @herzinger I'm actually with @florianbouvot on this one. I'd rather we offer sensible defaults (like for [widths fractions](https://github.com/inuitcss/inuitcss/blob/develop/utilities/_utilities.widths.scss#L45) for example), and have this map as an example in a comment.

**Comment by herzinger** [comment]
I'm okay with that, although as I said, nothing would be REALLY sensible defaults for this. 

That being said, if we follow the `x:y` convention it would also increase backward compatibility. I'll do this, and write some more detailed documentation for now.

File: objects/_objects.crop.scss [hunk 2]
```diff
@@ -4,12 +4,23 @@
 
 // A list of cropping ratios that get generated as modifier classes.
 // You should predefine it with only the ratios and names your project needs.
+//
+// the map keys are the strings used in the generated class names, and they can
```
Comment (csshugs) [inline objects/_objects.crop.scss]: Sorry for nitpicking, but can you please start the sentence with titlecase.

File: objects/_objects.ratio.scss [hunk 3]
```diff
@@ -4,12 +4,23 @@
 
 // A list of aspect ratios that get generated as modifier classes.
 // You should predefine it with only the ratios and names your project needs.
+//
+// the map keys are the strings used in the generated class names, and they can
```
Comment (csshugs) [inline objects/_objects.ratio.scss]: Same as for the crop object: Can you please start the sentence with titlecase.

**Review by csshugs — CHANGES_REQUESTED** [review]
This is IMO a very good compromise of providing a sensible default map with a good documented example of other possible class names.

If you could please fix the two typos I annotated, I'd be happy to merge this.

**Comment by herzinger** [comment]
Sure, no problem. I agree. 

In exchange, let me nitpick your nitpicking: what you mean is proper case. Title case is Something Like This. Also, you should say "start the sentence with an _uppercase_", or "_write_ the sentence _in_ proper case".

**Comment by dennisfrank** [comment]
C’mon guys. :wink: 

Most of the contributers aren’t native English speakers. Aiming for a uniform documentation/code language style is fine. But is it really necessary to criticize linguistic subtleties of discussion comments? 

(I’m just a (mostly) silent observer hoping for a v6 release soon with motivated contributors and core team! :smiley:)

**Comment by herzinger** [comment]
> But is it really necessary to criticize linguistic subtleties of discussion comments?

No, it's not. This was humor. Meta humor. Thats why I added the `xD"` at the end. Like how I only find it funny because I _literally_ am high on acid right now.

And now you are considering in your head if a guy that would do this kind of nitpicking would have used "literally" _figuratively_ in the sentence above. 

Like that.

`xD"`

**Review by csshugs — APPROVED** [review]

**Review by florianbouvot — APPROVED** [review]

**Comment by Jamiewarb** [comment]
This was an enjoyable read, culminating in a fine solution 👍

**Pull request merged**